Change privilege of enrich stats API to monitor (#52027)

ywangd opened the pull request:
The remote_monitoring_user user needs to access the enrich stats API.
But the request is denied because the API is categorized under admin.
The correct privilege should be monitor.

Changing action name breaks backwards compability as the name is serialised
and sent across the wire. Changes are made to send the old names for nodes
of old versions (before v7.7.0) and use new names for newer versions. This helps
v8.0 to not worry about any breaking changes and just use the new name.

relates: #51677

ywangd committed:
Add tests for changed action privileges

ywangd committed:
Update tests for better names

ywangd committed:
Add bwc for old action names

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authorization)

ywangd committed:
fix spotless error

ywangd commented on server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:
```diff
@@ -138,6 +138,12 @@ protected void resolveRequest(NodesRequest request, ClusterState clusterState) {
+     */
+    protected String getBwcTransportNodeAction(DiscoveryNode node) {
+        return transportNodeAction;
+    }
```
Opt-in for using an old action name.

ywangd commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/TransportEnrichStatsAction.java:
```diff
@@ -108,4 +110,9 @@ protected void masterOperation(
+    @Override
+    protected String getMasterActionName(DiscoveryNode node) {
+        return node.getVersion().before(Version.V_7_7_0) ? EnrichStatsAction.OLD_NAME : actionName;
+    }
```
Override existing methods for opt-in bwc action name.

ywangd commented:
Took a close look at the yaml tests and here is my understanding:
* If we are to add some bwc tests for the name changes, they would be part of `/rest-api-spec/[api,test]`. Both API and test spec files need to be added since there is nothing related to enrich yet. 
* The tests will be picked up by `/qa/mixed-cluster` which is part of the bwc tests
* Above tests have security disabled. So they will not test the permission changes. However, given the purpose of the new tests is to ensure name change do not break API calls and name resolution happens before security kicks in, I feel it is sufficient to just have them without security.
* To have ideal test coverage, the API should be made to both 7.7 and 7.6 node and they in turn should have a master node of different version, (e.g. `Request_1->7.6->7.7`, `Request_2->7.7->7.6`, ...).  But I am not sure how the yaml test suite handles this internally.
* There is a `mixed-cluster` section in the gradle file of `/x-pack/qa/multi-cluster-tests-with-security`. But I am not sure whether this is related.

ywangd committed:
Add smoke test for enrich stats API for mixed cluster

ywangd committed:
Fix bwc and add tests

ywangd commented:
Fixing bwc turned out to be trickly. The bwc method for action name, e.g. [getMasterNodeActionName](https://github.com/elastic/elasticsearch/blob/v7.6.0/server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java#L231), only solves the problem for one direction, i.e. from node of newer version to older version. However, when the request is send from node of older version, the message still contains the old action name. Hence [handler resolution](https://github.com/elastic/elasticsearch/blob/v7.6.0/server/src/main/java/org/elasticsearch/transport/InboundHandler.java#L165) still fails.

Therefore the current solution register both handler under both old and new names for handling requests from older versions. It is a bit messy but I cannot think of an alternative right now. On upside, it helps avoiding pollute 8.0.

ywangd committed:
Fix spotless

tvernum commented on server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:
```diff
@@ -138,6 +138,12 @@ protected void resolveRequest(NodesRequest request, ClusterState clusterState) {
+    /**
+     * Get a backwards compatible transport action name
+     */
+    protected String getBwcTransportNodeAction(DiscoveryNode node) {
```
I think this just should be 
```suggestion
    protected String getTransportNodeAction(DiscoveryNode node) {
```
I don't think the "BWC" helps with the method name - it assumes that this is only relevant for BWC, but really it's "what is the right action for this node" ?

tvernum replied:
Also, I think this part of the change should go to master - I know you don't need it for BWC purposes there, but we shouldn't make core API changes on 7.x only.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/enrich/action/EnrichStatsAction.java:
```diff
@@ -24,7 +24,8 @@
     public static final EnrichStatsAction INSTANCE = new EnrichStatsAction();
-    public static final String NAME = "cluster:admin/xpack/enrich/stats";
+    public static final String NAME = "cluster:monitor/xpack/enrich/stats";
+    public static final String BWC_NAME = "cluster:admin/xpack/enrich/stats";
```
Can you add a javadoc here that it had this name in pre 7.8 versions

ywangd replied:
Done

tvernum commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorStatsAction.java:
```diff
@@ -161,6 +184,11 @@ protected NodeResponse nodeOperation(NodeRequest request) {
+
+        @Override
+        protected String getBwcTransportNodeAction(DiscoveryNode node) {
+            return node.getVersion().before(Version.V_7_7_0) ? BWC_NAME + "[n]" : transportNodeAction;
```
This needs to update now that it's slipped.

ywangd replied:
Updated to V_7_9_0

ywangd committed:
Address feedback

ywangd committed:
spotless

ywangd committed:
Update version selector for yaml tests

ywangd commented:
@elasticmachine run elasticsearch-ci/1

ywangd committed:
Change privilege of enrich stats API to monitor (#52027)

The remote_monitoring_user user needs to access the enrich stats API.
But the request is denied because the API is categorized under admin.
The correct privilege should be monitor.

ycombinator commented:
Any idea when this PR might be merged and backported? Users who are using Metricbeat for Stack Monitoring have been getting 403 errors since 7.5.2.

ywangd commented:
> Any idea when this PR might be merged and backported? Users who are using Metricbeat for Stack Monitoring have been getting 403 errors since 7.5.2.

The current label is 7.9. We could also make it available in 7.8.1 as well.

ycombinator commented:
@ywangd That would be great. Given planned release dates, maybe 7.7.2 as well?

ywangd commented:
> @ywangd That would be great. Given planned release dates, maybe 7.7.2 as well?

Sorry I spoke too fast. We may not be able to port it to even `v7.8.1`. This is a breaking change in transport layer, which requires the older version, in this case `v7.9.0` to handle both old and new actions, i.e. `v7.9.0` becomes the shim. This type of change is difficult to be backported to multiple versions, i.e. combinatory explosion.

ywangd committed:
Minor tweak

ywangd committed:
Add bwc name javadoc

ywangd commented:
@tvernum I changed this PR into a backport. As discussed before, the code changes are a bit more than a regular PR because of bwc. I'd appreciate if you could give +1 before merging. Thanks.

ywangd committed:
Merge remote-tracking branch 'origin/7.x' into es-51677-enrich-stats-privilege-7.x

ywangd commented:
@elasticmachine run elasticsearch-ci/2

tvernum reviewed (approved):
LGTM

tvernum commented on x-pack/qa/rolling-upgrade/src/test/resources/rest-api-spec/test/mixed_cluster/110_enrich.yml:
```diff
@@ -0,0 +1,13 @@
+---
+"Enrich stats query smoke test for mixed cluster":
+  - skip:
+      version: " - 7.99.99"
```
Actually, is this right? This is the 7.9 backport.

ywangd replied:
A silly error of cherry pick and merge... This file (`110_enrich.yml`) is redundant and the correct one is `120_enrich.yml`. I replaced it with the content of `120_enrich.yml`.

ywangd committed:
Remove redundant file

ywangd committed:
Merge remote-tracking branch 'origin/7.x' into es-51677-enrich-stats-privilege-7.x

ywangd committed:
Merge remote-tracking branch 'origin/7.x' into es-51677-enrich-stats-privilege-7.x
